### PR TITLE
Environment brightness, contrast, saturation restore with 3d LUT.

### LINKED ISSUE #41257: Vulkan: Adjustments (brightness/contrast/saturation) not reimplemented yet
**Godot version:** Git https://github.com/godotengine/godot/commit/aded76cb84fd1081279d2ed11fd73b553c6e6b4e

**OS/device including version:** Fedora 31, GeForce GTX 1080 (NVIDIA 440.100)

**Issue description:**

After enabling adjustments in Environment, tweaking brightness/contrast/saturation does nothing.

**Steps to reproduce:**

Enable Adjustments in Environment and tweak the values.

**Minimal reproduction project:** [test_adjustments.zip](https://github.com/godotengine/godot/files/5075212/test_adjustments.zip)

### PR DESCRIPTION
**Pull request by fire**
Co-authored-by: CptPotato
Co-authored-by: QbieShay

Restored brightness, contrast, saturation adjustments.

*Bugsquad edit: This closes https://github.com/godotengine/godot/issues/41257.*

See also https://github.com/godotengine/godot-proposals/issues/1382

### CODE REVIEW

File: servers/rendering/rasterizer_rd/shaders/tonemap.glsl [hunk 1]
```diff
@@ -251,7 +251,11 @@ vec3 apply_bcs(vec3 color, vec3 bcs) {
 }
 
 vec3 apply_color_correction(vec3 color, sampler3D correction_tex) {
-	return texture(correction_tex, color).rgb;
+	ivec3 tex_size = textureSize(correction_tex, 0);
+	color = texture(correction_tex,
+			color * (vec3(tex_size - ivec3(1)) / vec3(tex_size)) + vec3(0.5) / vec3(tex_size))
+					.rgb;
+	return color;
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/shaders/tonemap.glsl:258]: That's not very readable. Maybe instead:
```suggestion
	vec3 coord = color * (vec3(tex_size - ivec3(1)) / vec3(tex_size)) + vec3(0.5) / vec3(tex_size);
	return texture(correction_text, coord).rgb;
```

And further for the actual equation:
```
	vec3 coord = (color * vec3(tex_size - ivec3(1)) + vec3(0.5)) / vec3(tex_size);
```
Should also be faster.

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h [hunk 2]
```diff
@@ -640,13 +640,15 @@ class RasterizerEffectsRD {
 		float auto_exposure_grey = 0.5;
 		RID exposure_texture;
 
+		bool use_adjustments = false;
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:643]: This is assigned but never used it seems?

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h [hunk 3]
```diff
@@ -640,13 +640,15 @@ class RasterizerEffectsRD {
...
+
 		bool use_bcs = false;
 		float brightness = 1.0;
 		float contrast = 1.0;
 		float saturation = 1.0;
 
 		bool use_color_correction = false;
-		RID color_correction_texture;
+		RID color_correction_texture = RID();
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:651]: Isn't `RID` default initialized already?

File: servers/rendering/rasterizer_rd/shaders/tonemap.glsl (hunk 1 above)
Comment (fire) [reply servers/rendering/rasterizer_rd/shaders/tonemap.glsl:258]: This equation doesn't work. I had trouble and wanted someone else to look over this part.

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h (hunk 2 above)
Comment (fire) [reply servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:643]: I thought this was used. Let me check.

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h (hunk 3 above)
Comment (fire) [reply servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:651]: Will remove.

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h (hunk 2 above)
Comment (fire) [reply servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:643]: This is used in:

```c++
// rasterizer_scene_rd.cpp
		if (can_use_effects && env) {
			tonemap.use_adjustments = env->adjustments_enabled;
```

Comment (akien-mga) [reply servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:643]: Well that's not use, that's assignment, which is what I commented :)

If it's assigned but never actually used as a condition, it seems to be unused.

File: servers/rendering/rasterizer_rd/shaders/tonemap.glsl (hunk 1 above)
Comment (fire) [reply servers/rendering/rasterizer_rd/shaders/tonemap.glsl]: The color correction is still broken, but updated code.

**Comment by fire** [comment]
Test project for future iFire.

[Env Adjustments Game Project.zip](https://github.com/godotengine/godot/files/5372240/Env.Adjustments.Game.Project.zip)

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h (hunk 2 above)
Comment (fire) [reply servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:643]: Let me see what happens when I delete it.

Comment (fire) [reply servers/rendering/rasterizer_rd/rasterizer_effects_rd.h]: Deleted.

File: servers/rendering/rasterizer_rd/shaders/tonemap.glsl [hunk 4]
```diff
@@ -251,7 +251,9 @@ vec3 apply_bcs(vec3 color, vec3 bcs) {
 }
 
 vec3 apply_color_correction(vec3 color, sampler3D correction_tex) {
-	return texture(correction_tex, color).rgb;
+	ivec3 tex_size = textureSize(correction_tex, 0);
```
Comment (reduz) [inline servers/rendering/rasterizer_rd/shaders/tonemap.glsl:254]: I suggest if possible pass the texture size in the push constant, because textureSize is akin to a texture read.

**Comment by fire** [comment]
Setting the Color correction new sets the screen grey and still doesn't work. Revised with Reduz's comments.

**Comment by fire** [comment]
Will rebase.

**Comment by fire** [comment]
There's another patch incoming.

**Comment by fire** [comment]
Screen Shot 2020-11-11 at 4 24 42 PM

Sunset lut!

File: servers/rendering/rasterizer_rd/rasterizer_scene_rd.h [hunk 5]
```diff
@@ -1571,7 +1579,16 @@ class RasterizerSceneRD : public RasterizerScene {
 	RS::EnvironmentSSRRoughnessQuality environment_get_ssr_roughness_quality() const;
 
 	void environment_set_tonemap(RID p_env, RS::EnvironmentToneMapper p_tone_mapper, float p_exposure, float p_white, bool p_auto_exposure, float p_min_luminance, float p_max_luminance, float p_auto_exp_speed, float p_auto_exp_scale);
-	void environment_set_adjustment(RID p_env, bool p_enable, float p_brightness, float p_contrast, float p_saturation, RID p_ramp) {}
+	void environment_set_adjustment(RID p_env, bool p_enable, float p_brightness, float p_contrast, float p_saturation, RID p_ramp) {
+		Environment *env = environment_owner.getornull(p_env);
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/rasterizer_scene_rd.h:1583]: Function should be in the cpp file

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.cpp [hunk 6]
```diff
@@ -738,7 +738,7 @@ void RasterizerEffectsRD::tonemapper(RID p_source_color, RID p_dst_framebuffer,
 	RD::get_singleton()->draw_list_bind_uniform_set(draw_list, _get_uniform_set_from_texture(p_source_color), 0);
 	RD::get_singleton()->draw_list_bind_uniform_set(draw_list, _get_uniform_set_from_texture(p_settings.exposure_texture), 1);
 	RD::get_singleton()->draw_list_bind_uniform_set(draw_list, _get_uniform_set_from_texture(p_settings.glow_texture, true), 2);
-	RD::get_singleton()->draw_list_bind_uniform_set(draw_list, _get_uniform_set_from_texture(p_settings.color_correction_texture), 3);
+	RD::get_singleton()->draw_list_bind_uniform_set(draw_list, _get_uniform_set_from_texture(p_settings.color_correction_texture, true), 3);
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/rasterizer_effects_rd.cpp:741]: Why is this using mip maps?

File: servers/rendering/rasterizer_rd/shaders/tonemap.glsl [hunk 7]
```diff
@@ -257,7 +257,8 @@ vec3 apply_bcs(vec3 color, vec3 bcs) {
 }
 
 vec3 apply_color_correction(vec3 color, sampler3D correction_tex) {
-	return texture(correction_tex, color).rgb;
+	vec3 uv_interp = color;
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/shaders/tonemap.glsl:260]: We usually use uv_interp for the uv that comes from the vertex shader. I think we can just eliminate this line altogether.

**Review by clayjohn — CHANGES_REQUESTED** [review]

File: modules/mono/glue/GodotSharp/GodotSharp/Core/GD.cs [hunk 8]
```diff
@@ -39,6 +39,14 @@ public static real_t DecTime(real_t value, real_t amount, real_t step)
...
         }
 
+        public static FuncRef FuncRef(Object instance, StringName funcName)
+        {
+            var ret = new FuncRef();
+            ret.SetInstance(instance);
+            ret.SetFunction(funcName);
+            return ret;
+        }
+
```
Comment (aaronfranke) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/GD.cs:49]:
```suggestion
```

**Review by clayjohn — COMMENTED** [review]
Code looks fine now. 

My only holdback is the change in workflow/behaviour. This change is going to stop people from being able to import their Environments into 4.0. Also, it makes the workflow way trickier as Godot doesn't have any 3D texture authoring tools. 

I think we need to have the ability to use a texture2D or a texture3D. That way users can continue to use their current Environments while having the option of using a more complex 3D LUT.

**Comment by fire** [comment]
I though the design was to use 3D luts in master and I wasn't sure of a design that allows both.

**Comment by Madaraz1024** [comment]
LUT only texture3D and does not support texture2D

lut already experiment the result well --> https://youtu.be/FacL-08WpP8 

already tried the gimp editor handle colors well import godot texture3D --> https://youtu.be/4lYdYqshSmY

**Comment by fire** [comment]
I don't think it's the responsibility of this pr to write this conversion. I imagined that there was a way to convert Texture2D into a Texture3D using default settings of 8 tiles width and 8 tiles height. 

A restart is required to convert a Texture2D to a Texture3D. A Texture2D that is 2D can be converted to a Texture3D of 64x64x1 which is 1 Tile width and 1 Tile depth compared to a regular Texture3D that is a 2D image converted to 64x64x64. Then, bi-linear interpolation from the 3D Material should function. I don't have any examples of a 2D LUT to test.

**Comment by clayjohn** [comment]
The current workflow for users is that they can create a LUT using a single gradient texture. While it is clear that a 1D LUT is extremely limiting, its what every Godot game uses and it is very easy for users to get something working. They just add a gradient texture and can have results in seconds. 

Switching to a 3D LUT breaks every game that is using the 1D LUT and it makes adding a LUT considerably more difficult as the LUT needs to be generated outside of the engine. 

While I understand that the 3D LUT is much more powerful and expressive (and we need support for them), there has been no discussion about breaking the previous workflow, in particular, there has been no proposal weighing the pros and cons. 

That being said, adding support for texture2Ds and texture3Ds would be easy and would avoid creating breaking changes, so that is what I prefer. Otherwise, we need to find out from users if they are okay losing the current LUT approach.

**Comment by fire** [comment]
The workflow for conversion of a 2D LUT would be to turn a png into a 1 tile width and 1 tile height 3d texture, which is doable.

On a related note, I had a hard time finding anyone providing examples for the 2D LUT workflow in Godot Engine.

Did you mean 2D LUT? A 1D LUT is a line. I think you can interpolate a line into a cube.

**Comment by clayjohn** [comment]
The current approach in (3.2) accepts a Texture2D which represents a 1D LUT. 

I am not suggesting a conversion tool (although a conversion tool is a viable idea). I am suggested allowing users to use the current workflow and the new 3D texture workflow.

**Comment by Madaraz1024** [comment]
1D LUT example below screenshot? right

image
RGB Table 16х1 size

**Comment by fire** [comment]
He means this. 
2d-gradient-3

**Comment by fire** [comment]
Example in 3.2

image

Screen Shot 2020-11-12 at 8 11 26 AM

As you can see here an 1D LUT can be a 3D LUT when you set tile width and tile height to 1.

**Comment by Madaraz1024** [comment]
I'm think 1D LUT better use a gradient no textures needed :)

**Comment by clayjohn** [comment]
@fire I guess the important thing is just allowing users to continue using GradientTextures. So, you can convert it to a 3D Texture under the hood if you want, or you can adapt it so the shader accepts either.

**Comment by fire** [comment]
editor_screenshot_2020-11-13T210147-0800

Got the Gradient texture to dynamically change. Needs testing.

**Comment by Madaraz1024** [comment]
good no problem
image

**Comment by fire** [comment]
For others searching for what happens if you have a 3D LUT that is kept as a texture2D., the result looks spectacular.

The original is a 2D texture and it's supposed to be sliced 8 weight and 8 height into a 3D texture.

Screen Shot 2020-11-13 at 9 33 12 PM

9d9nqmzl0hf21

We discussed this and we decided the result is just a misuse of the feature, but we can't distinguish between the 1D and 3D LUT encoded into the same png.

File: scene/resources/environment.cpp [hunk 9]
```diff
@@ -891,23 +894,58 @@ float Environment::get_adjustment_saturation() const {
...
 }
 
 void Environment::_update_adjustment() {
+	Ref<GradientTexture> grad_tex = adjustment_color_correction;
+	adjustment_color_correction_cache = adjustment_color_correction;
+	Ref<Texture2D> texture_2d = adjustment_color_correction_cache;
+	Ref<ImageTexture3D> texture_3d_image = adjustment_color_correction_cache;
+	Ref<StreamTexture3D> texture_3d_stream = adjustment_color_correction_cache;
+	RID color_correction = adjustment_color_correction_cache.is_valid() ? adjustment_color_correction_cache->get_rid() : RID();
+	if (texture_2d.is_valid()) {
```
Comment (clayjohn) [inline scene/resources/environment.cpp:913]: You are reloading the texture every time ``update_adjustment()`` is called. This means if you use any of the sliders, you may be loading the 3D texture thousands of times.

Comment (fire) [reply scene/resources/environment.cpp:913]: Is delaying the load by 0.3 of a second a good alternative?

Comment (fire) [reply scene/resources/environment.cpp:913]: This technique was done for screenshots to avoid capturing menus.

Comment (clayjohn) [reply scene/resources/environment.cpp:913]: Delaying isn't a good idea, again because this function is called when sliders are used for BCS.

Comment (fire) [reply scene/resources/environment.cpp:913]: Any suggestions? I'm out of ideas.

Comment (fire) [reply scene/resources/environment.cpp]: @clayjohn Mentioned using different shader versions. Still unresolved, but we can retest with the new options.

image

**Comment by Madaraz1024** [comment]
LUT 3D 16x1 will future replace the old 8x8 ?
image

File: servers/rendering/rasterizer_rd/shaders/tonemap.glsl [hunk 10]
```diff
@@ -23,7 +23,8 @@ layout(location = 0) in vec2 uv_interp;
 layout(set = 0, binding = 0) uniform sampler2D source_color;
 layout(set = 1, binding = 0) uniform sampler2D source_auto_exposure;
 layout(set = 2, binding = 0) uniform sampler2D source_glow;
-layout(set = 3, binding = 0) uniform sampler3D color_correction;
+layout(set = 3, binding = 0) uniform sampler2D color_correction_1d;
+layout(set = 4, binding = 0) uniform sampler3D color_correction_3d;
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/shaders/tonemap.glsl:27]: Mobile devices can only have 4 uniform sets. 

The easiest way for this to work is to use a new shader version as we discussed on IRC. A shader version is created by adding an ``ifdef`` statement. 

If you want I can make a PR to your branch, just let me know.

**Comment by fire** [comment]
Go ahead.

On Nov 22, 2020, at 12:52 AM, Clay John <notifications@github.com> wrote:

﻿
@clayjohn commented on this pull request.

In servers/rendering/rasterizer_rd/shaders/tonemap.glsl:

> @@ -23,7 +23,8 @@ layout(location = 0) in vec2 uv_interp;
 layout(set = 0, binding = 0) uniform sampler2D source_color;
 layout(set = 1, binding = 0) uniform sampler2D source_auto_exposure;
 layout(set = 2, binding = 0) uniform sampler2D source_glow;
-layout(set = 3, binding = 0) uniform sampler3D color_correction;
+layout(set = 3, binding = 0) uniform sampler2D color_correction_1d;
+layout(set = 4, binding = 0) uniform sampler3D color_correction_3d;
Mobile devices can only have 4 uniform sets.

The easiest way for this to work is to use a new shader version as we discussed on IRC. A shader version is created by adding an ifdef statement.

If you want I can make a PR to your branch, just let me know.

**Comment by fire** [comment]
Used ClayJohn's revision.

```C++
void Environment::set_adjustment_color_correction(Ref<Texture> p_color_correction) {
	adjustment_color_correction = p_color_correction;
	Ref<GradientTexture> grad_tex = p_color_correction;
	if (grad_tex.is_valid()) {
		if (!grad_tex->is_connected(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment))) {
			grad_tex->connect(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment));
			if (grad_tex->get_gradient().is_valid() && !grad_tex->get_gradient()->is_connected(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment))) {
				grad_tex->get_gradient()->connect(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment));
			}
		}
	}
	Ref<Texture2D> adjustment_texture_2d = adjustment_color_correction;
	if (adjustment_texture_2d.is_valid()) {
		use_1d_color_correction = true;
	} else {
		use_1d_color_correction = false;
	}
	_update_adjustment();
}
```

Also restored hard-coded gradient texture updating support. Probably want to make a note that if there's no hard coding, an editor restart is needed.

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h [hunk 11]
```diff
@@ -181,7 +183,7 @@ class RasterizerEffectsRD {
 
 		uint32_t glow_texture_size[2];
 		float glow_intensity;
-		uint32_t pad3;
+		uint32_t use_1d_ramp;
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:186]: Should still be pad3

File: scene/resources/texture.cpp [hunk 12]
```diff
@@ -832,11 +832,17 @@ void ImageTexture3D::_update(const TypedArray<Image> &p_data) {
 }
 
 Error ImageTexture3D::create(Image::Format p_format, int p_width, int p_height, int p_depth, bool p_mipmaps, const Vector<Ref<Image>> &p_data) {
-	RID tex = RenderingServer::get_singleton()->texture_3d_create(p_format, p_width, p_height, p_depth, p_mipmaps, p_data);
+	RID tex = RS::get_singleton()->texture_3d_create(p_format, p_width, p_height, p_depth, p_mipmaps, p_data);
```
Comment (clayjohn) [inline scene/resources/texture.cpp:835]: Why is this change included?

File: scene/resources/environment.cpp [hunk 13]
```diff
@@ -891,23 +894,41 @@ float Environment::get_adjustment_saturation() const {
 	return adjustment_saturation;
 }
 
-void Environment::set_adjustment_color_correction(const Ref<Texture2D> &p_ramp) {
-	adjustment_color_correction = p_ramp;
+void Environment::set_adjustment_color_correction(Ref<Texture> p_color_correction) {
+	adjustment_color_correction = p_color_correction;
+	Ref<GradientTexture> grad_tex = p_color_correction;
+	if (grad_tex.is_valid()) {
```
Comment (clayjohn) [inline scene/resources/environment.cpp:900]: This stuff isn't necessary. Since we just pass the RID to the rendering device, all updates happen automatically.

File: scene/resources/environment.cpp [hunk 14]
```diff
@@ -31,6 +31,9 @@
 #include "environment.h"
 
 #include "core/config/project_settings.h"
+#include "core/core_string_names.h"
+#include "core/error/error_macros.h"
```
Comment (clayjohn) [inline scene/resources/environment.cpp:35]: Are the additional includes needed?

**Review by clayjohn — CHANGES_REQUESTED** [review]

File: scene/resources/environment.cpp (hunk 13 above)
Comment (fire) [reply scene/resources/environment.cpp:900]: But gradient changing doesn't work. I guess we can document that you need to restart the editor to modify the graident..

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h (hunk 11 above)
Comment (fire) [reply servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:186]: Mismerge.

File: scene/resources/texture.cpp (hunk 12 above)
Comment (fire) [reply scene/resources/texture.cpp:835]: No necessary. Reverting.

File: scene/resources/environment.cpp (hunk 14 above)
Comment (fire) [reply scene/resources/environment.cpp]: These are used for set_adjustment_color_correction's gradient signals.

File: scene/resources/environment.cpp (hunk 13 above)
Comment (fire) [reply scene/resources/environment.cpp:898]: Closing, please reopen if you have arguments why we need to restart the editor to change colours on the gradient colours list.

Comment (clayjohn) [reply scene/resources/environment.cpp:898]: Did you test it out? You can update the gradient and everything updates automatically. There is no point in sending the same RID to the server over and over again. 

Take a look at the 3.2 code, it worked the same way.

Comment (fire) [reply scene/resources/environment.cpp:898]: Yes I did. I can give you a video.

Comment (fire) [reply scene/resources/environment.cpp:898]: https://imgur.com/a/gsVy0zq Here are the two variations of the changes. One has this code removed.

Comment (fire) [reply scene/resources/environment.cpp:898]: The only one that works is when I connect the gradient changed signal to changing the adjustments.

File: scene/resources/environment.cpp [hunk 15]
```diff
@@ -891,23 +892,41 @@ float Environment::get_adjustment_saturation() const {
...
+	adjustment_color_correction = p_color_correction;
+	Ref<GradientTexture> grad_tex = p_color_correction;
+	if (grad_tex.is_valid()) {
+		if (!grad_tex->is_connected(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment))) {
+			grad_tex->connect(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment));
+			if (grad_tex->get_gradient().is_valid() && !grad_tex->get_gradient()->is_connected(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment))) {
+				grad_tex->get_gradient()->connect(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment));
+			}
+		}
+	}
```
Comment (clayjohn) [inline scene/resources/environment.cpp:905]:
```suggestion
	if (grad_tex.is_valid()) {
		if (!grad_tex->is_connected(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment))) {
			grad_tex->connect(CoreStringNames::get_singleton()->changed, callable_mp(this, &Environment::_update_adjustment));
		}
	}
```

You are right. The RID needs to be updated when the GradientTexture changes. But it doesn't need to be updated everytime the Gradient changes.

File: scene/resources/texture.cpp [hunk 16]
```diff
@@ -834,9 +834,15 @@ void ImageTexture3D::_update(const TypedArray<Image> &p_data) {
 Error ImageTexture3D::create(Image::Format p_format, int p_width, int p_height, int p_depth, bool p_mipmaps, const Vector<Ref<Image>> &p_data) {
 	RID tex = RenderingServer::get_singleton()->texture_3d_create(p_format, p_width, p_height, p_depth, p_mipmaps, p_data);
 	ERR_FAIL_COND_V(tex.is_null(), ERR_CANT_CREATE);
-
+	format = p_format;
```
Comment (clayjohn) [inline scene/resources/texture.cpp:837]: These imageTexture changes seem to have snuck back in.

Comment (fire) [reply scene/resources/texture.cpp:837]: I believe you did not remove them in your pr.

Comment (clayjohn) [reply scene/resources/texture.cpp:837]: I didn't touch texture.cpp in my PR. But this PR doesn't require these changes, so they shouldn't be included. 

Everything else looks good to go though!

Comment (fire) [reply scene/resources/texture.cpp]: Removed.

**Review by clayjohn — APPROVED** [review]
Woohoo! Great work!

**Comment by reduz** [comment]
Looks ok to me

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Note that the "Co-authored-by:" fields for @CptPotato and @QbieShay don't seem to work, I think you need to include their email (check their recently used name + email in `git log`).

**Comment by fire** [comment]
@akien-mga Done.

**Comment by akien-mga** [comment]
Thanks!
